lib/rust/ensogl/core/src/display/scene.rs

```diff
@@ -164,39 +180,71 @@ impl Mouse {
                     let position_bottom_left = Vector2(new_pos.x as f32, new_pos.y as f32);
                     let position_top_left = Vector2(new_pos.x as f32, shape.height - new_pos.y as f32);
                     let position = position_bottom_left - shape.center();
-                    frp.position_bottom_left.emit(position_bottom_left);
-                    frp.position_top_left.emit(position_top_left);
-                    frp.position.emit(position);
+                    frp_deprecated.position_bottom_left.emit(position_bottom_left);
+                    frp_deprecated.position_top_left.emit(position_top_left);
+                    frp_deprecated.position.emit(position);
+
+                    pointer_target_registry.with_mouse_target(target.get(), |_, d| {
+                        d.emit_event(event.clone());
+                    });
                 }
             }),
-        ));
-        let on_down = mouse_manager.on_down.add(js_event.handler(
-            f!([frp, click_count, display_mode] (event:&mouse::OnDown) {
+        );
+        let on_down = mouse_manager.on_down.add(
+            f!([pointer_target_registry, target, last_pressed_elem, frp_deprecated, click_count, display_mode]
+            (event:&mouse::Down) {
                 click_count.modify(|v| *v += 1);
                 if display_mode.get().allow_mouse_events() {
-                    frp.down.emit(event.button());
+                    let button = event.button();
+                    frp_deprecated.down.emit(button);
+
+                    let current_target = target.get();
+                    last_pressed_elem.borrow_mut().insert(button, current_target);
+                    pointer_target_registry.with_mouse_target(current_target, |t, d| {
+                        t.emit_mouse_down(button);
+                        d.emit_event(event.clone());
+                    });
                 }
             }),
-        ));
-        let on_up = mouse_manager.on_up.add(js_event.handler(
-            f!([frp, display_mode] (event:&mouse::OnUp) {
+        );
+        let on_up = mouse_manager.on_up.add(
+            f!([pointer_target_registry, target, last_pressed_elem, frp_deprecated, display_mode]
+            (event: &mouse::Up) {
                 if display_mode.get().allow_mouse_events() {
-                    frp.up.emit(event.button())
+                    let button = event.button();
+                    frp_deprecated.up.emit(button);
+
+                    let current_target = target.get();
+                    if let Some(last_target) = last_pressed_elem.borrow_mut().remove(&button) {
+                        pointer_target_registry.with_mouse_target(last_target, |t, d| {
+                            t.emit_mouse_release(button);
+                            d.emit_event(event.clone().unchecked_convert_to::<mouse::Release>());
```

Comment:
Add comment why doing so unchecked is safe.

Reply:
As agreed via phone, we are OK with it as it has doc comments.